File: #1.data-structures/#2.Stack/src/kotlin/modules/UndoRedo.kt
class UndoRedo<T> {
  private var undoStack = Stack<T>()
  private var redoStack = Stack<T>()

  fun addAction(action: T) {
      undoStack.push(action)
      redoStack = Stack<T>() // Limpa a pilha de redo
  }

  fun undo(): T? {
      val action = undoStack.pop()
      if (action != null) {
          redoStack.push(action)
      }
      return action
  }

  fun redo(): T? {
      val action = redoStack.pop()
      if (action != null) {
          undoStack.push(action)
      }
      return action
  }

  fun debug() {
      println("Undo Stack: ${undoStack.getItems()}")
      println("Redo Stack: ${redoStack.getItems()}")
  }
}

File: #1.data-structures/#2.Stack/src/typescript/src/modules/UndoRedo.ts
import { Stack } from "./Stack";

export class UndoRedo<T> {
  private undoStack: Stack<T> = new Stack<T>();
  private redoStack: Stack<T> = new Stack<T>();

  addAction(action: T): void {
    this.undoStack.push(action);
    this.redoStack = new Stack(); // Limpa a pilha de redo
  }

  undo(): T | undefined {
    const action = this.undoStack.pop();
    if (action !== undefined) {
      this.redoStack.push(action);
    }

    return action;
  }

  redo() {
    const action = this.redoStack.pop();
    if (action !== undefined) {
      this.undoStack.push(action);
    }

    return action;
  }

  debug(): void {
    console.log("Undo Stack: ", [...this.undoStack.getItems()]);
    console.log("Redo Stack: ", [...this.redoStack.getItems()]);
  }
}
